### PR TITLE
1072: chartjs integration

### LINKED ISSUE #1072: New charts engine
@roiLeo has raised awareness of the big bundle of current chart

Proposed to use https://www.chartjs.org/ 
If we manage to keep original functionality, I'm up to it! 

Mind making PR against `main-nuxt` branch

### PR DESCRIPTION
**Pull request by rhaicode**
**Thank you for your contribution** to the KodaDot NFT gallery.

👇 Let's do a quick check before the merge.

### PR type
- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### Before submitting Pull Request, please make sure:
- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted screenshot of demonstrated change in this PR

### Optional
- [ ] I've tested PR on mobile and everything works
- [ ] I found edge cases

### What's new?
- [x] PR closes #1072 
- [x]  Replaced charts using ECharts to ChartJS instead

### Had issue bounty label?
- [x] Fill up your KSM address: [Payout](https://kodadot.xyz/transfer/?target=EJ4LJsZAefUW5pwokk5G5b96CpHaABiXXaEUpdBakP8oYJr)

### Community participation
- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot
- [x] My fix has changed **something** on UI, a screenshot for others, is best to understand changes.

### CODE REVIEW

**Comment by rhaicode** [comment]
Graph found at: /en/rmrk/gallery/:id

Before (ECharts):
before-echarts

After (ChartJS with chartjs-adapter-luxon):
after-using-chartjs

There are some follow-up questions I want to ask:
1.) Are we going to follow 24 hour format for the x-axis?
2.) Are we going to support data zooming and panning?

Thanks a lot!

**Comment by yangwao** [comment]
Amazing progress! 

1.) Are we going to follow the 24-hour format for the x-axis? 

Thinking we should with our ethos, least possible information, that means having. 
Let's keep 24 hours, just HH if possible have precise HH: MM at event point.
 
2.) Are we going to support data zooming and panning?

That's a tricky one, how hard it would be and how easy would be interacting with that on mobile?

I would propose
- item detail -- w/o zoom/pan - I'm thinking to add pre-made time frames. 7d/14d/1m where 7d could be the default
- collection history -- w/ zoom/pan -- as there is a lot of data

**Comment by rhaicode** [comment]
1.) All right, got it. I'll be working on it right away.
2.) Unlike ECharts that already supports this by providing a `dataZoom` property, ChartJS doesn't support it by default. We could add a plugin like [chartjs-plugin-zoom](https://www.chartjs.org/chartjs-plugin-zoom/samples/basic.html)

**Comment by yangwao** [comment]
> 2.) Unlike ECharts that already supports this by providing a `dataZoom` property, ChartJS doesn't support it by default. We could add a plugin like [chartjs-plugin-zoom](https://www.chartjs.org/chartjs-plugin-zoom/samples/basic.html)

Works well on mobile, let's try using it!

**Comment by yangwao** [comment]
Chatted with @rhaicode and will migrate collection chart to `main-nuxt` as well

**Comment by rhaicode** [comment]
Hello @yangwao , before I proceed with migrating the collection chart to `main-nuxt`, I'd like to ask for your feedback with the chart having zoom plugin?

https://screencast-o-matic.com/watch/crXIezVI11T

**Comment by rhaicode** [comment]
> I would propose
> 
> item detail -- w/o zoom/pan - I'm thinking to add pre-made time frames. 7d/14d/1m where 7d could be the default

Currently working on this.

> collection history -- w/ zoom/pan -- as there is a lot of data

In order to work on the graph under the Activity tab which is under the Collections tab, the Collections functionality should be added as well since `main-nuxt` doesn't have the Collections tab above.

**Comment by yangwao** [comment]
> Activity tab which is under the Collections tab, the Collections functionality should be added as well since `main-nuxt` doesn't have the Collections tab above.

Yes, let's add it :)

**Comment by yangwao** [comment]
> Hello @yangwao. I'd like to ask for your feedback with the chart having zoom plugin?
> 
> https://screencast-o-matic.com/watch/crXIezVI11T

Is there possibility to fix Y axis not go below 0 ? as price will be always >= 0

**Comment by yangwao** [comment]
Added test case
- https://deploy-preview-1110--koda-nuxt.netlify.app/rmrk/gallery/7479313-7a147cba01afccbf17-GA-GENERATIVE_ASTERISM_23-0000000000000023
- https://deploy-preview-1110--koda-nuxt.netlify.app/rmrk/gallery/7479313-7a147cba01afccbf17-GA-GENERATIVE_ASTERISM_346-0000000000000346
- https://deploy-preview-1110--koda-nuxt.netlify.app/rmrk/gallery/7479313-7a147cba01afccbf17-GA-GENERATIVE_ASTERISM_342-0000000000000342

**Comment by yangwao** [comment]
Stradford pinged me on Thursday will check on this one

**Comment by rhaicode** [comment]
Hello, @yangwao ! This is the link for the possible v1 for the chartjs integration. 
http://somup.com/crXbQl0QE8

**Comment by yangwao** [comment]
For the collection view, let's make it 100% width. 

image

For item detail, let's add colours. 
Green for BUY, List can be as it is.

image

I sense that zoom in is proportional, but zoom out is not symmetric?

https://user-images.githubusercontent.com/5887929/142200691-8695350d-e81b-44db-9580-06ef8b55a5aa.mov

**Comment by rhaicode** [comment]
> For the collection view, let's make it 100% width.
Got this.

> For item detail, let's add colours.
> Green for BUY, List can be as it is.
Does this mean the color of the dots on the price chart?

> I sense that zoom in is proportional, but zoom out is not symmetric?
I have only set the minimum range for zoom out to be 0 KSM (so that we won't be able to see negative KSM values). However, I'd like to ask for some clarity regarding this.
- What do you mean by proportional and symmetric within this context? I'm only using the default zooming of chartjs-plugin-zoom

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 1]
```diff
@@ -1,160 +1,196 @@
...
+    document!.getElementById('collectionPriceChart')!.style.cursor = 'grabbing';
+  }
+
+  protected onCanvasMouseUp() {
+    document!.getElementById('collectionPriceChart')!.style.cursor = 'auto';
+  }
+
+  protected onCanvasMouseLeave() {
+    document!.getElementById('collectionPriceChart')!.style.cursor = 'auto';
+  }
```
Comment (roiLeo) [inline components/rmrk/Gallery/CollectionPriceChart.vue:58]: missing return type `void`

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 2]
```diff
@@ -1,160 +1,196 @@
...
+      const chart = new Chart(ctx, {
+        type: 'line',
+        data: {
+          datasets: [
+            {
+              label: 'Floor Price',
+              data: this.priceData[0].map((data) => ({
+                x: data[0],
+                y: data[1],
+              })),
```
Comment (roiLeo) [inline components/rmrk/Gallery/CollectionPriceChart.vue:82]: missing `(data)` type

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 3]
```diff
@@ -1,160 +1,196 @@
...
+              pointBorderColor: 'blue',
+              pointRadius: 4,
+              pointHoverRadius: 6,
+            },
+            {
+              label: 'Sold NFT Price',
+              data: this.priceData[1].map((data) => ({
+                x: data[0],
+                y: data[1],
+              })),
```
Comment (roiLeo) [inline components/rmrk/Gallery/CollectionPriceChart.vue:95]: missing `(data)` type

File: components/rmrk/Gallery/PriceChart.vue [hunk 4]
```diff
@@ -1,155 +1,169 @@
...
+      const ctx = (
+        document?.getElementById('priceChart') as HTMLCanvasElement
+      )?.getContext('2d')!;
+      const chart = new Chart(ctx, {
+        type: 'line',
+        data: {
+          datasets: [
+            {
+              label: 'Price',
+              data: this.priceData.map((data) => ({ x: data[0], y: data[1] })),
```
Comment (roiLeo) [inline components/rmrk/Gallery/PriceChart.vue:57]: missing `(data)` type

**Review by roiLeo — COMMENTED** [review]
Hey! 
I would love to see what it looks like but I can't run nuxt branch right now.
Why do we have 2 different Charts component? Different feature/data?

**Comment by rhaicode** [comment]
@roiLeo hello! yes, we have two since i just pull from `main-nuxt` and removed ECharts there. I'll be working on the requested code changes.

**Comment by yangwao** [comment]
> > For item detail, let's add colours.
> > Green for BUY, List can be as it is.
> 
> * Does this mean the color of the dots on the price chart?

Let's keep it the same as a graph for collection activity with lines :)


> Why do we have 2 different Charts component? Different feature/data?

@roiLeo we will discard the old chart engine for this one which should be more native and works much better on mobile.

**Comment by rhaicode** [comment]
@roiLeo the next commit now provides the requested code changes.

@yangwao for the chart at `en/rmrk/gallery/:id`, it seems that the `priceData` that is only pushed to the chart follows this condition:
```
if (event['Type'] === 'SET-PRICE' || event['Type'] === 'CREATE') {
  this.priceData.push([date, this.formatPrice(event['Amount'])])
}
```
Since the collection chart has two lines, the floor price and the sold nft price, does this mean we are going to have two lines as well for the chart at `en/rmrk/gallery/:id`, like buy price and list price?

**Comment by yangwao** [comment]
> @yangwao for the chart at `en/rmrk/gallery/:id`, it seems that the `priceData` that is only pushed to the chart follows this condition:
> 
> ```
> if (event['Type'] === 'SET-PRICE' || event['Type'] === 'CREATE') {
>   this.priceData.push([date, this.formatPrice(event['Amount'])])
> }
> ```
> 
> Since the collection chart has two lines, the floor price and the sold nft price, does this mean we are going to have two lines as well for the chart at `en/rmrk/gallery/:id`, like buy price and list price?

Yes, that's the case. Two lines, LIST and BUY.  Check `main` branch how was recently history chart refactored by @vikiival most probably it will end up in `main-nuxt` branch as well

**Comment by rhaicode** [comment]
Latest chart for `/en/rmrk/gallery/:id`
http://somup.com/crXrlU0oz5

**Comment by yangwao** [comment]
> but then it seems that the endpoint for the charts at collections>activity tab is not working? i'm only receiving no data in my end

I guess needs to have more data there, cc @dezine2dev

**Comment by yangwao** [comment]
> Latest chart for `/en/rmrk/gallery/:id` http://somup.com/crXrlU0oz5

How does it look like on NFT which has more interactions?

**Comment by rhaicode** [comment]
@yangwao Something like this? http://somup.com/crX3n80DQE

**Comment by yangwao** [comment]
> @yangwao Something like this? http://somup.com/crX3n80DQE

okay, doable I guess! 

I guess we are waiting for the graph for collections and it's everything done? 

Check recent merge https://github.com/kodadot/nft-gallery/pull/1212 and refactor for charts was made in `main`, in best case migrate that code and I'm happy to top your bounty for this issue!

**Comment by yangwao** [comment]
Beautiful!
For me it's mergeable @roiLeo @vikiival  

https://deploy-preview-1110--koda-nuxt.netlify.app/en/rmrk/collection/4a5668af20c28c074f-KSMDUCKZ

image

https://deploy-preview-1110--koda-nuxt.netlify.app/en/rmrk/collection/800f8a914281765a7d-KITTY
image

**Comment by rhaicode** [comment]
@yangwao The code for both charts come from `main` too then just migrated it here to `main-nuxt`. @roiLeo could also verify

**Comment by yangwao** [comment]
> Hey! I would love to see what it looks like but I can't run nuxt branch right now. 

@roiLeo  https://deploy-preview-1110--koda-nuxt.netlify.app/en/rmrk/collection/4a5668af20c28c074f-KSMDUCKZ

**Comment by yangwao** [comment]
> > Hey! I would love to see what it looks like but I can't run nuxt branch right now.
> 
> @roiLeo https://deploy-preview-1110--koda-nuxt.netlify.app/en/rmrk/collection/4a5668af20c28c074f-KSMDUCKZ

@roiLeo let's merge it then if it's good!

**Comment by yangwao** [comment]
payout $400 sent meanwhile! https://kusama.subscan.io/extrinsic/0x5945e757597ec9c581749d5bad66c52ba3b3148fef35ca60a3801e1f98c3f093

**Comment by rhaicode** [comment]
received! thanks a lot 🚀

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 5]
```diff
@@ -1,160 +1,180 @@
...
   <div class="price-chart mt-4">
-    <p class="label">
-      {{ $t('Price Chart') }}
-    </p>
-    <div id="chart" ref="chart" class="echart"></div>
+    <div class="is-flex is-align-items-center is-justify-content-space-between">
+      <p class="label">
+        {{ $t('Price Chart') }}
+      </p>
+      <button class="button" @click="resetZoom">Reset zoom</button>
```
Comment (roiLeo) [inline components/rmrk/Gallery/CollectionPriceChart.vue:7]:
```suggestion
      <b-button type="is-primary" @click="resetZoom">Reset zoom</b-button>
```
I would change it by buefy button as I got weird behavior while button is active and :focus

**Review by roiLeo — APPROVED** [review]
Collection charts LGTM, I'm not able to test GalleryItem chart as route not working for now.
We could increase chart height on mobile view (167px is kinda small to me)

File: components/rmrk/Gallery/CollectionPriceChart.vue (hunk 5 above)
Comment (rhaicode) [reply components/rmrk/Gallery/CollectionPriceChart.vue:7]: Okay, I'm on it

**Comment by rhaicode** [comment]
I'll provide follow up commits. Thanks

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 6]
```diff
@@ -1,160 +1,180 @@
...
+                  enabled: true,
+                },
+                maxRotation: 0,
+                minRotation: 0,
+                color: '#fff',
+              },
+            },
+            y: {
+              ticks: {
+                callback: (value) => {
```
Comment (vikiival) [inline components/rmrk/Gallery/CollectionPriceChart.vue:156]: I would take the unit from the vuex `store`

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 7]
```diff
@@ -1,160 +1,180 @@
...
+              borderColor: '#d32e79',
+              tension: 0.3,
+              pointBackgroundColor: 'white',
+              pointBorderColor: 'blue',
+              pointRadius: 4,
+              pointHoverRadius: 6,
+            },
+            {
+              label: 'Sold NFT Price',
+              data: this.priceData[1].map((data: PriceDataArray) => ({
```
Comment (vikiival) [inline components/rmrk/Gallery/CollectionPriceChart.vue:93]: I would move this that arrow function into the utils

**Review by vikiival — COMMENTED** [review]
Looks legit.
also use linter

File: components/rmrk/Gallery/CollectionPriceChart.vue (hunk 6 above)
Comment (rhaicode) [reply components/rmrk/Gallery/CollectionPriceChart.vue:156]: Hello, I'm not sure about what you are referring to as the unit from the vuex store. The y-axis of the chart will always be in KSM. I also can't find the unit to use for the ticks at y-axis from the vuex store.

File: components/rmrk/Gallery/CollectionPriceChart.vue (hunk 7 above)
Comment (rhaicode) [reply components/rmrk/Gallery/CollectionPriceChart.vue]: On it

**Comment by rhaicode** [comment]
@roiLeo @vikiival  this one's ready for review

**Comment by vikiival** [comment]
Deff check tomorrow !

File: components/rmrk/Gallery/CollectionPriceChart.vue [hunk 8]
```diff
@@ -1,160 +1,175 @@
...
 
   protected chartOptionsLine: any = {};
-  protected Chart!: ECharts.ECharts;
+  protected Chart!: Chart<'line', any, unknown>;
+
+  protected resetZoom(): void {
+    this.Chart.resetZoom();
+  }
 
   protected onWindowResize() {
```
Comment (vikiival) [inline components/rmrk/Gallery/CollectionPriceChart.vue:44]: I would personally debounce this :D

File: components/rmrk/Gallery/CollectionPriceChart.vue (hunk 6 above)
Comment (vikiival) [reply components/rmrk/Gallery/CollectionPriceChart.vue:151]: > The y-axis of the chart will always be in KSM

Only applies only for the Kusama. 

We will run KodaDot on [Basilisk](https://github.com/galacticcouncil/Basilisk-node) which has BSX token

File: components/rmrk/Gallery/History.vue [hunk 9]
```diff
@@ -1,168 +1,217 @@
...
       if (!('To' in event)) {
-        event['To'] = newEvent['caller']
-        prevOwner = event['To']
+        event['To'] = newEvent['caller'];
+        prevOwner = event['To'];
       }
 
       // Amount
-      event['Amount'] = Vue.filter('formatBalance')(curPrice, 12, 'KSM')
+      event['Amount'] = formatBalance(curPrice, this.decimals, this.unit);
```
Comment (vikiival) [inline components/rmrk/Gallery/History.vue:173]: 👁️ I see some correct usage of unit

File: components/rmrk/Gallery/PriceChart.vue [hunk 10]
```diff
@@ -1,155 +1,153 @@
...
+                },
+                maxRotation: 0,
+                minRotation: 0,
+                color: '#fff',
+              },
+            },
+            y: {
+              ticks: {
+                callback: (value) => {
+                  return `${Number(value).toFixed(2)} KSM`;
```
Comment (vikiival) [inline components/rmrk/Gallery/PriceChart.vue:131]: Missing Unit here

**Comment by vikiival** [comment]
Build is failing because of #458

**Comment by rhaicode** [comment]
@vikiival implemented requested code changes

**Review by vikiival — APPROVED** [review]

**Comment by yangwao** [comment]
Amazing job @vikiival @roiLeo @rhaicode 
looking forward this stuff live soon(tm) ! :)

**Pull request merged**